components/script/dom/bindings/root.rs

```diff
@@ -47,21 +47,87 @@ use std::ptr;
 use std::rc::Rc;
 use style::thread_state;
 
-/// A rooted reference to a DOM object.
+/// A rooted value.
 ///
 /// The JS value is pinned for the duration of this object's lifetime; roots
 /// are additive, so this object's destruction will not invalidate other roots
 /// for the same JS value. `Root`s cannot outlive the associated
 /// `RootCollection` object.
 #[allow(unrooted_must_root)]
 #[allow_unrooted_interior]
-pub struct DomRoot<T: DomObject> {
-    /// Reference to rooted value that must not outlive this container
-    ptr: Dom<T>,
+pub struct Root<T: StableTraceObject> {
+    /// The value to root.
+    value: T,
     /// List that ensures correct dynamic root ordering
     root_list: *const RootCollection,
 }
 
+impl<T> Root<T>
+where
+    T: StableTraceObject + 'static,
+{
+    /// Create a new stack-bounded root for the provided value.
+    /// It cannot outlive its associated `RootCollection`, and it gives
+    /// out references which cannot outlive this new `Root`.
+    #[allow(unrooted_must_root)]
+    unsafe fn new(value: T) -> Self {
+        debug_assert!(thread_state::get().is_script());
+        STACK_ROOTS.with(|ref root_list| {
+            let root_list = &*root_list.get().unwrap();
+            root_list.root(value.stable_trace_object());
+            Root { value, root_list }
+        })
+    }
+}
+
+pub unsafe trait StableTraceObject {
+    fn stable_trace_object(&self) -> *const JSTraceable;
+}
+
+unsafe impl<T> StableTraceObject for Dom<T>
+where
+    T: DomObject,
+{
+    fn stable_trace_object<'a>(&'a self) -> *const JSTraceable {
+        #[allow(unrooted_must_root)]
+        struct ReflectorStackRoot(Reflector);
+        unsafe impl JSTraceable for ReflectorStackRoot {
+            unsafe fn trace(&self, tracer: *mut JSTracer) {
+                trace_reflector(tracer, "on stack", &self.0);
+            }
+        }
+        unsafe {
+            &*(self.reflector() as *const Reflector as *const ReflectorStackRoot)
```

Comment:
I would prefer to dereference the `*const Reflector` and get the ReflectorStackRoot field instead of casting the pointer twice.

Reply:
What do you mean? There is no `ReflectorStackRoot` field in `Reflector`, it's the opposite.